### PR TITLE
Feature: Distribution Packages

### LINKED ISSUE #96: Debian packaging for Slurm exporter
Greetings 👋

[I made a Debian package for this exporter](https://github.com/NucciTheBoss/prometheus-slurm-exporter/tree/deb-pkg/debian) and I wanted to contribute it back to the repository. I just had a couple questions before opening the PR:

1. Where would you like for me to place the Debian package recipe in the source tree? Currently I just have a top-level `debian/` directory since `build/` is taken by `just` for the built exporter binaries.

2. How necessary are the SWIG bindings for the exporter? The reason I ask is because they were causing the Debian package build to break as some of the data structures were undefined. I was able to fix the issue by removing `cmain.go` and `cext` from the source tree before executing `dpkg-buildpackage` and building the version that uses the Slurm CLI commands. I am using [`dh_golang`](https://go-team.pages.debian.net/packaging.html#_use_dh_golang) to compile the exporter for the Debian package. The tool seems to not like the presence of both `cmain.go` and `main.go` within the source tree. 

    Here's the errors that I was seeing when trying to build the exporter with `cmain.go` and `cext` still in the source tree:
    
    ```text
    github.com/rivosinc/prometheus-slurm-exporter/cext
    # github.com/rivosinc/prometheus-slurm-exporter/cext
    src/github.com/rivosinc/prometheus-slurm-exporter/cext/server.go:16:63: undefined: Destructor
    src/github.com/rivosinc/prometheus-slurm-exporter/cext/server.go:22:18: undefined: NewNodeFetcher
    src/github.com/rivosinc/prometheus-slurm-exporter/cext/server.go:25:17: undefined: NewJobFetcher
    src/github.com/rivosinc/prometheus-slurm-exporter/cext/server.go:29:31: undefined: Destructor
    dh_auto_build: error: cd obj-x86_64-linux-gnu && go install -trimpath -v -p 1 github.com/rivosinc/prometheus-slurm-exporter 
    github.com/rivosinc/prometheus-slurm-exporter/cext github.com/rivosinc/prometheus-slurm-exporter/exporter returned exit code 1
    make: *** [debian/rules:7: binary] Error 255
    dpkg-buildpackage: error: debian/rules binary subprocess returned exit status 2
    debuild: fatal error at line 1184:
    dpkg-buildpackage -us -uc -ui --build=binary --build=binary failed
    ```

I'm also currently working on setting up a PPA in Launchpad for hosting public builds of the exporter for Ubuntu. Hope to chat soon!

### PR DESCRIPTION
**Pull request by drewstinnett**
This PR adds RPM and Deb packaging from goreleaser, as well as a little systemd file to start and stop the service.

resolves #96

### CODE REVIEW

**Comment by drewstinnett** [comment]
Oops, need to fix reuse workflow..

**Comment by NucciTheBoss** [comment]
Hah, looks like we had the same idea this week (#96) - this is great! :laughing: 

I published a [PPA for Noble and Jammy on Launchpad](https://launchpad.net/~ubuntu-hpc/+archive/ubuntu/prometheus-slurm-exporter) for the exporter. I've reached out to the Debian Go team about getting the exporter uploaded to Debian. It would be great to have latest builds published to GitHub and a `git buildpackage` version on Salsa for inclusion into Debian :smile:

**Comment by drewstinnett** [comment]
> Hah, looks like we had the same idea this week (#96) - this is great! 😆
> 
> I published a [PPA for Noble and Jammy on Launchpad](https://launchpad.net/~ubuntu-hpc/+archive/ubuntu/prometheus-slurm-exporter) for the exporter. I've reached out to the Debian Go team about getting the exporter uploaded to Debian. It would be great to have latest builds published to GitHub and a `git buildpackage` version on Salsa for inclusion into Debian 😄

Oh I totally missed that...good timing!!

File: .goreleaser.yaml [hunk 1]
```diff
@@ -10,3 +10,24 @@ builds:
       - amd64
       - arm
       - arm64
+nfpms:
+  -
+    vendor: rivosinc
+    maintainer: abhinavDhulipal
```
Comment (abhinavDhulipala) [inline .goreleaser.yaml]: :)

```suggestion
    maintainer: abhinavDhulipala
```

**Comment by abhinavDhulipala** [comment]
Howdy! First of all thanks a ton for your contributions. Give me a day or 2 wrap my mind around this MR and get back to you. In the meanwhile, I don't _think_ tracing functionality is supported the way it is packaged right now. Please checkout my latest MR #95 for how to package and resolve the html files to support tracing. In short, we need to package the `./exporter/templates/*` html files. In the systemd file and pkgs, we need to support pointing to the proper templated directory using the [TRACE_ROOT_PATH](https://github.com/rivosinc/prometheus-slurm-exporter?tab=readme-ov-file#exporter-env-var-docs) env var.

**Comment by drewstinnett** [comment]
Ah yes that makes sense I'll add that in shortly... Thank you!

**Comment by drewstinnett** [comment]
Ok @abhinavDhulipala, I think the packaging is correct now if you'd like to take a look when you can (No rush here btw, and thank you so much for this awesome project! 🙏)

Just a thought, but how would you feel if I changed the TRACE_ROOT_PATH logic in the code to look in the order:

1. $TRACE_ROOT_PATH path
2. . (cwd)
3. /usr/share/prometheus-slurm-exporter/templates/

I don't think that would break any current workflows, but it would allow users of the distro packages to use tracing without having to specify an additional env variable. If that is something you would be interested in, I'd be happy to make that change as well as part of this PR. No worries if you'd like to keep it the same though too.

Thank you again!

**Comment by abhinavDhulipala** [comment]
Howdy @drewstinnett that trace path feature looks like a great feature, but we could also configure the environment variable of the systemd file to fill in the $TRACE_PATH to `/usr/share/prometheus-slurm-exporter/templates`. If we were to implement the scheme above, I think we should implement it in the following way:

1. If `TRACE_ROOT_PATH` is specified, search that directory. If we don't find a `templates` dir, let's panic and crash the program.
2. If `TRACE_ROOT_PATH` **isn't** specified, we can search cwd and /usr/share/prometheus-slurm-exporter. 


Now, I tried to install debian found on your forks release [page](https://github.com/drewstinnett/prometheus-slurm-exporter/releases/tag/v0.0.3). I did the following in case you want to repro:

```bash
docker run -ti --rm --platform linux/amd64 ubuntu:22.04 bash
$ apt update -y && apt-get install wget -y && mkdir /workspace
$ cd /workspace
$ wget https://github.com/drewstinnett/prometheus-slurm-exporter/releases/download/v0.0.3/prometheus-slurm-exporter_0.0.3_linux_amd64.deb
$  apt-get install -f prometheus-slurm-exporter_0.0.3_linux_amd64.deb
# unfortunately got the following
$  stat /usr/share/prometheus-slurm-exporter/         
stat: cannot statx '/usr/share/prometheus-slurm-exporter/': No such file or directory
```

It doesn't look like the templates were packaged. Is this stale perhaps?

Thanks again for all your hard work!

**Comment by drewstinnett** [comment]
Thank you @abhinavDhulipala, I will work on these fixes/changes!

**Comment by drewstinnett** [comment]
Working on the trace logic...quick question for you @abhinavDhulipala, should the app do a 'deep' search of the TRACE_ROOT_PATH for the 'templates' directory, or do we always expect it to be right under the top path? For example:

if TRACE_ROOT_PATH is set to "/srv", should "/srv/subdir/templates" be considered valid, or would we only want to look at "/srv/templates"? Either is fine with me, just wanted to check the intended behavior. Thank you!

**Comment by abhinavDhulipala** [comment]
I have it now as a recursive search, but I think that could be a potential bottleneck if configured wrong. Let's modify it to simply checking for the existence of the one html file we have now.

**Comment by drewstinnett** [comment]
Ok, whenever you get a chance, can you give it another review? I think I was mis-using the github-release command, which appears to have generated the packages based off of an older tag. I think this is fixed now. I verified with:

```bash
$ goreleaser release --clean --snapshot
...
$ dpkg-deb -c dist/prometheus-slurm-exporter_1.6.5-SNAPSHOT-0969bb7_linux_amd64.deb
drwxr-xr-x root/root         0 2024-09-03 10:02 ./usr/
drwxr-xr-x root/root         0 2024-09-03 10:02 ./usr/bin/
-rwxr-xr-x root/root  10014872 2024-09-03 10:02 ./usr/bin/prometheus-slurm-exporter
drwxr-xr-x root/root         0 2024-09-03 10:02 ./usr/lib/
drwxr-xr-x root/root         0 2024-09-03 10:02 ./usr/lib/systemd/
drwxr-xr-x root/root         0 2024-09-03 10:02 ./usr/lib/systemd/system/
-rw-r--r-- root/root       250 2024-09-03 09:35 ./usr/lib/systemd/system/prometheus-slurm-exporter.service
drwxr-xr-x root/root         0 2024-09-03 10:02 ./usr/share/
drwxr-xr-x root/root         0 2024-09-03 10:02 ./usr/share/prometheus-slurm-exporter/
drwxr-xr-x root/root         0 2024-09-03 10:02 ./usr/share/prometheus-slurm-exporter/templates/
-rw-r--r-- root/root       927 2024-08-27 10:47 ./usr/share/prometheus-slurm-exporter/templates/proc_traces.html
```

The logic for the detection is also in place now

```bash
$ LOGLEVEL=debug prometheus-slurm-exporter -trace.enabled=true
time=2024-09-03T12:22:40.135-04:00 level=INFO msg="trace path enabled at path: :9092/trace"
time=2024-09-03T12:22:40.136-04:00 level=DEBUG msg="using trace template path: /usr/share/prometheus-slurm-exporter/templates"
time=2024-09-03T12:22:40.136-04:00 level=INFO msg="serving metrics at :9092/metrics"
…
$ LOGLEVEL=debug TRACE_ROOT_PATH=/tmp/empty-dir prometheus-slurm-exporter -trace.enabled=true
time=2024-09-03T12:23:46.371-04:00 level=INFO msg="trace path enabled at path: :9092/trace"
panic: TRACE_ROOT_PATH must include a directory called: templates
…
```

Happy to make whatever tweaks/changes you suggest in addition to these as well. Thank you!

File: packaging/postinstall.sh [hunk 2]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+# SPDX-FileCopyrightText: 2023 Rivos Inc.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+systemctl daemon-reload
+systemctl enable --now prometheus-slurm-exporter.service
```
Comment (abhinavDhulipala) [inline packaging/postinstall.sh]: I'm conflicted on whether we should install a systemctl function by default. I personally won't be using this method of deployment and when I try to develop and test out of containers this gets in the way quite a bit. I'd suggest remove this entirely. We could guard with the following: `[[ -d /run/systemd/system ]] && ...`, but I still would prefer to drop this and just put this as documentation. Perhaps a curlable snippet of some kind.

File: exporter/trace_test.go [hunk 3]
```diff
@@ -202,3 +205,28 @@ func TestPython3Wrapper(t *testing.T) {
...
+	os.Clearenv()
+	testDir := t.TempDir()
+	t.Setenv("TRACE_ROOT_PATH", testDir)
+	// Ensure that the function panics if given a TRACE_ROOT_PATh with no 'templates' subdirectory
+	assert.PanicsWithValue(t, "TRACE_ROOT_PATH must include a directory called: templates", func() { detectTraceTemplatePath() })
+	require.NoError(t, os.Mkdir(filepath.Join(testDir, templateDirName), 0o700))
+
+	// Now that we have a 'templates' subdir, it should no longer panic
+	assert.Equal(t, filepath.Join(testDir, templateDirName), detectTraceTemplatePath())
+}
```
Comment (abhinavDhulipala) [inline exporter/trace_test.go:219]: This is an awesome test! Great work.

**Review by abhinavDhulipala — CHANGES_REQUESTED** [review]
The biggest takeaway here is that I've found it cumbersome to install this in containers which I often use for deployment. Let me know what you think!

EDIT: I've also validated this within a docker container to ensure that your branch works :).

File: packaging/postinstall.sh (hunk 2 above)
Comment (drewstinnett) [reply packaging/postinstall.sh]: I understand what you mean...I'll make this update removing the systemd components from the packaging and include a snippet in the documentation that can be copy/pasted. Thank you for that feedback!

Comment (drewstinnett) [reply packaging/postinstall.sh]: Ok, systemd bits have been removed from the packaging, and I added a little copy/pastable snippet to the README for folks that would like to enable it as a systemd service. Look ok? Thank you!

**Review by abhinavDhulipala — APPROVED** [review]
This looks awesome, thanks a ton for all your hard work once again. I'll cut a release on main shortly

**Pull request merged**